### PR TITLE
Added transfer timeout for sectionfilter data.

### PR DESCRIPTION
**Pull request by pipelka**
This is the first draft for the new transfer logic.
I haven't tested it so far on my production system. Will do that tomorrow.
I'm not quite sure about the EAGAIN case but I think it's the right way to resend the packet.

### CODE REVIEW

File: sectionfilter.c [hunk 1]
```diff
@@ -212,25 +212,39 @@ void cSatipSectionFilter::Process(const uint8_t* dataP)
...
   if (section) {
      uchar *data = section->Data();
      int count = section->Count();
      if (data && (count > 0) && (socketM[1] >= 0) && (socketM[0] >= 0)) {
-        ssize_t len = send(socketM[1], data, count, MSG_EOR);
-        ERROR_IF(len < 0 && errno != EAGAIN, "send()");
-        if (len > 0) {
-           ringBufferM->Drop(section);
-           result = !!ringBufferM->Available();
+        if (TransferData(socketM[1], data, count, 2000)) {
```
Comment (rofafor) [inline sectionfilter.c]: The timeout is way too big as it blocks data transfers to all other filters as well. Any polling logic should be implemented into the cSatipSectionFilterHandler and keep all filter file handles in it at the same time.

**Comment by pipelka** [comment]
The timeout value was just a shot in the dark. 10ms seems to be reasonable.

Polling could be done with all available descriptors at once and then send packets to all "writeable" ones.
I think you thought of this kind of behaviour.

**Comment by pipelka** [comment]
2nd round.
Using all filedescriptors for polling now. Looks better to me (no more ringbuffer overflows).

**Comment by pipelka** [comment]
I also modified the cSatipSectionFilterHandler::Action() loop. It now tries to process all pending data and then sends all the resulting section data.
Section data is currently not resent on failure and the error message will also change in the final version.

File: sectionfilter.c [hunk 2]
```diff
@@ -270,54 +271,68 @@ cSatipSectionFilterHandler::~cSatipSectionFilterHandler()
...
-                  if (filtersM[i])
-                     filtersM[i]->Process(p);
-                  }
-              mutexM.Unlock();
-              processed = true;
-              continue;
-              }
-           }
-        cCondWait::SleepMs(10); // to avoid busy loop and reduce cpu load
+        //cCondWait::SleepMs(10); // to avoid busy loop and reduce cpu load
```
Comment (rofafor) [inline sectionfilter.c]: We should sleep - at least if there weren't anything to send.

File: sectionfilter.c [hunk 3]
```diff
@@ -270,54 +271,68 @@ cSatipSectionFilterHandler::~cSatipSectionFilterHandler()
...
+         }
+      }
+
+  // anyone ready for writing
+  if (poll(pollFdsM, eMaxSecFilterCount, 10) <= 0)
+     return false;
+
+  // send data
+  for (unsigned int i = 0; i < eMaxSecFilterCount; ++i) {
+      if(pollFdsM[i].revents & POLLOUT)
```
Comment (rofafor) [inline sectionfilter.c]: Missing whitespace after 'if'

File: sectionfilter.c [hunk 4]
```diff
@@ -270,54 +271,68 @@ cSatipSectionFilterHandler::~cSatipSectionFilterHandler()
...
+  // assemble all handlers to poll
+  for (unsigned int i = 0; i < eMaxSecFilterCount; ++i) {
+      if (filtersM[i] && filtersM[i]->Available() != 0) {
+         pollFdsM[i].fd = filtersM[i]->GetFd();
+         pollFdsM[i].events = POLLOUT;
+         }
+      }
+
+  // anyone ready for writing
+  if (poll(pollFdsM, eMaxSecFilterCount, 10) <= 0)
```
Comment (rofafor) [inline sectionfilter.c]: Define the timeout as enum in the class: ```eSecFilterSendTimeoutMs = 10```

File: sectionfilter.c [hunk 5]
```diff
@@ -212,27 +212,28 @@ void cSatipSectionFilter::Process(const uint8_t* dataP)
...
-        if (len > 0) {
-           ringBufferM->Drop(section);
-           result = !!ringBufferM->Available();
+        if (send(socketM[1], data, count, MSG_EOR) > 0) {
            // Update statistics
-           AddSectionStatistic(len, 1);
+           AddSectionStatistic(count, 1);
            }
+        else
+          esyslog("failed to send section data (%i bytes) to fd: %i (errno: %i)", count, socketM[1], errno);
```
Comment (rofafor) [inline sectionfilter.c]: Use ```error()``` (or ERROR_IF) macro. Include ```deviceIndexM``` information as elsewhere, if using ```error()```.

File: sectionfilter.c (hunk 2 above)
Comment (pipelka) [reply sectionfilter.c]: AFAIK, cRingBuffer::Get() blocks if the buffer is empty. I think sleeping here is unneccessary. I do not see any difference in CPU usage with of without the sleep. But maybe I'm wrong.

File: sectionfilter.c (hunk 3 above)
Comment (pipelka) [reply sectionfilter.c]: It's not my natural style of coding. But I'm getting better, ...

File: sectionfilter.c (hunk 2 above)
Comment (rofafor) [reply sectionfilter.c]: Oh, I forgot that there are 100ms timeouts defined in the constructor for the ring buffer. Please, remove the dead code.

File: sectionfilter.c (hunk 3 above)
Comment (rofafor) [reply sectionfilter.c]: I'm trying to keep the way of KLS here (except for parameter/member naming).

**Comment by pipelka** [comment]
I have this version running for a couple of hours now on my development system and see sporadic errors in sending frames. Trying to resend these frames results in an infinite loop. Bit confused about that.
Is it possible that vdr closes it's end of the socket ?

**Comment by rofafor** [comment]
VDR is using OpenFIlter/CloseFilter interface, so you should see any socket handling stuff in plugin behaviour as well. It might be that VDR just can't keep up reading the socket from time to time. You can tweak that one by tweaking the kernel settings for unix socket size and ring buffer timeouts.

**Comment by pipelka** [comment]
Ok. Fixed the final bits. I'm unsure if it's ready to pull. I think it should settle a bit.
Some more testing should be done, ...

**Comment by pipelka** [comment]
Running for 24h on my production system.
No problems so far. CPU usage seems to be a bit lower.

**Comment by rofafor** [comment]
Great to hear! I'll merge the PR after I've validated the functionality on my setup too.

File: sectionfilter.c [hunk 6]
```diff
@@ -270,54 +271,67 @@ cSatipSectionFilterHandler::~cSatipSectionFilterHandler()
...
+                }
+
         // Send demuxed section packets through all filters
-        bool retry = false;
         mutexM.Lock();
-        for (unsigned int i = 0; i < eMaxSecFilterCount; ++i) {
-            if (filtersM[i] && filtersM[i]->Send())
-               retry = true;
-            }
+        while (Send()) ;
```
Comment (rofafor) [inline sectionfilter.c]: It might be better to include this "while" functionality inside the ```Send()``` method and rename it into e.g. ```SendAll()```.

**Comment by pipelka** [comment]
Here you are.

File: sectionfilter.c [hunk 7]
```diff
@@ -270,54 +271,68 @@ cSatipSectionFilterHandler::~cSatipSectionFilterHandler()
       Delete(i);
 }
 
+void cSatipSectionFilterHandler::SendAll(void)
+{
+  while (true) {
```
Comment (rofafor) [inline sectionfilter.c]: The infinite loop is a bad design habit. I think "do {} while (available)" would be pretty much perfect here or even "for (i=0; i<100;++i) break;" where number of maximum loops is a finite.

Comment (pipelka) [reply sectionfilter.c]: What about using
<pre>
while (Running()) {}
</pre>
it would also handle the case of infinite looping.

Comment (pipelka) [reply sectionfilter.c]: Hmm,... Thought a second time about it. I don't even see the case of infinite looping.

**Pull request merged**